main.py

```diff
@@ -359,80 +344,126 @@ def print_tree(tree, path, level):
+                f"{path}{pad}{passed:>5}/{count:<5} ({percentage:6.2f}%) {results} "
+            )
             if passed > 0:
                 for k, v in tree["children"].items():
                     print_tree(v, path + "/" + k, level + 1)
 
         for k, v in self.directory_result_map.items():
             print_tree(v, k, 0)
 
-    def process(self, file: Path) -> TestRun:
+    def process_list(self, file_list: list[Path]) -> List[TestRun]:
+        if len(file_list) == 0:
+            return []
+
         try:
-            return run_test(
+            return run_test_list(
                 self.libjs_test262_runner,
                 self.test262_root,
-                file,
+                [str(p) for p in file_list],
                 use_bytecode=self.use_bytecode,
+                parse_only=self.parse_only,
                 timeout=self.timeout,
                 memory_limit=self.memory_limit,
+                update_function=self.update_function,
             )
-        except:
-            return TestRun(
-                file,
-                result=TestResult.RUNNER_EXCEPTION,
-                output=traceback.format_exc(),
-                exit_code=None,
-                strict_mode=None,
-            )
+        except Exception as e:
+            return [
+                TestRun(
+                    str(file),
+                    result=TestResult.RUNNER_EXCEPTION
+                    if i == 0
+                    else TestResult.SKIPPED,
+                    output=traceback.format_exc() if i == 0 else "",
+                    exit_code=0,
+                    strict_mode=False,
+                )
+                for i, file in enumerate(file_list)
+            ]
 
     def run(self) -> None:
         if not self.files:
             self.log("No tests to run.")
             return
-        show_progress = not self.silent
-        if show_progress:
+
+        workers = self.concurrency
+
+        if self.shuffle:
+            random.shuffle(self.files)
+
+        amount_of_work_lists = workers
+        if self.total_count > workers * workers * 4:
+            amount_of_work_lists = workers * 4
+
+        if self.lists != -1:
+            amount_of_work_lists = self.lists
+
+        amount_of_work_lists = min(amount_of_work_lists, self.total_count)
+        work_lists = [[] for i in range(amount_of_work_lists)]
+        j = 0
+        for j, f in enumerate(self.files):
+            work_lists[j % amount_of_work_lists].append(f)
+
+        if not self.silent:
             progressbar = tqdm(
                 total=self.total_count, mininterval=1, unit="tests", smoothing=0.1
             )
+
+            def update_progress(value):
+                progress_mutex.acquire()
+                try:
+                    progressbar.update(value)
+                finally:
+                    progress_mutex.release()
+
+            self.update_function = update_progress
+
+            def write_output(message: str = ""):
```

Comment:
`write_output()` seems like a mistake, why the default argument?

Reply:
In the current runner we have to interrupt the progress bar to output failure/verbose tests.
But I found that it has a method for that `tqdm.write` only problem is that it does not have a default argument breaking `print()` which is why I added this wrapper method.